cli/upgrade/upgrade.go

```diff
@@ -0,0 +1,183 @@
+package upgrade
+
+import (
+	_ "embed"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/tarantool/tt/cli/connector"
+	"github.com/tarantool/tt/cli/replicaset"
+	"github.com/tarantool/tt/cli/running"
+)
+
+//go:embed lua/upgrade.lua
+var upgradeLuaScript string
+
+type SyncInfo struct {
+	LSN uint32  `mapstructure:"lsn"`
+	IID uint32  `mapstructure:"iid"`
+	Err *string `mapstructure:"err"`
+}
+
+func internalUpgrade(conn connector.Connector) (SyncInfo, error) {
+	var upgradeInfo SyncInfo
+	res, err := conn.Eval(upgradeLuaScript, []any{}, connector.RequestOpts{})
+	if err != nil {
+		return upgradeInfo, err
+	}
+	if err := mapstructure.Decode(res[0], &upgradeInfo); err != nil {
+		return upgradeInfo, err
+	}
+	if upgradeInfo.Err != nil {
+		return upgradeInfo, errors.New(*upgradeInfo.Err)
+	}
+	return upgradeInfo, nil
+}
+
+func WaitLSN(conn connector.Connector, masterUpgradeInfo SyncInfo, timeout int) error {
+	for i := 0; i < timeout; i++ {
+		var res []interface{}
+		res, err := conn.Eval(
+			fmt.Sprintf("return box.info.vclock[%d]", masterUpgradeInfo.IID),
+			[]any{}, connector.RequestOpts{})
+		if err != nil || len(res) == 0 {
+			return err
+		}
+		var lsn uint32
+		switch v := res[0].(type) {
+		case uint16:
+			lsn = uint32(v)
+		case uint32:
+			lsn = v
+		}
+		if lsn >= masterUpgradeInfo.LSN {
+			return nil
+		}
+		time.Sleep(time.Second)
+	}
+	return errors.New("")
```

Comment:
I'm a bit worrying about such an error. Can we write at least 'timeout exceeded' into it? Even if it is not used anywhere, it is more readable.

Reply:
Taking into account my suggestion below to report either last eval's error or the last seen LSN to a user, I think we should include the information about the last LSN into this message to concatenate it into the report printed to the user.

Reply:
Fixed.

Reply:
Not exactly: the last seen LSN is not reported to the user (if eval succeeds, but the `lsn >= masterLSN` check fails). I would set `lastError` in this case too.

Reply:
> I would set `lastError` in this case too.

Added.